Make decimal type fields support push down in kudu connector

Crossoverrr opened the pull request:
This pr can make decimal type fields support push down in kudu connector. 
otherwise when we using a decimal column in where clause will have error below:
![image](https://user-images.githubusercontent.com/33892550/69533455-d1ff8280-0fb2-11ea-9d9b-b5eaa1f94798.png)
ps: column "$sum_visit_day" is a decimal column.

Crossoverrr commented:
add package imports

Praveen2112 commented:
hi, Thanks a lot for raising this PR. Can you please squash the two commits into one. I guess you have submitted the CLA

Praveen2112 commented on presto-kudu/src/main/java/io/prestosql/plugin/kudu/KuduClientSession.java:
```diff
@@ -543,7 +543,11 @@ private KuduPredicate createComparisonPredicate(ColumnSchema columnSchema, KuduP
         Object javaValue = TypeHelper.getJavaValue(type, value);
         if (javaValue instanceof Long) {
-            return KuduPredicate.newComparisonPredicate(columnSchema, op, (Long) javaValue);
+            if (type.getClass().getSuperclass().equals(DecimalType.class)) {
```
Can we use `instanceof` here ? 
```suggestion
            if (type instanceof DecimalType) {
```

Crossoverrr replied:
You are right! I will do that!

Praveen2112 commented on presto-kudu/src/main/java/io/prestosql/plugin/kudu/KuduClientSession.java:
```diff
@@ -543,7 +543,11 @@ private KuduPredicate createComparisonPredicate(ColumnSchema columnSchema, KuduP
         Object javaValue = TypeHelper.getJavaValue(type, value);
         if (javaValue instanceof Long) {
-            return KuduPredicate.newComparisonPredicate(columnSchema, op, (Long) javaValue);
+            if (type.getClass().getSuperclass().equals(DecimalType.class)) {
```
What if we could handle in `TypeHelper` instead of here, In the line 158 https://github.com/prestosql/presto/blob/a6e7ab26195732cb3821b32443bf96a780d163a2/presto-kudu/src/main/java/io/prestosql/plugin/kudu/TypeHelper.java#L158

Crossoverrr replied:
That make sense, I can return a BigDecimal(nativeValue) in TypeHelper#getJavaValue when type instanceof DecimalType is trye, and at KuduClientSession#createComparisonPredicate I can handle this value directly.
You are right!

ebyhr commented on presto-kudu/src/main/java/io/prestosql/plugin/kudu/KuduClientSession.java:
```diff
@@ -543,7 +544,11 @@ private KuduPredicate createComparisonPredicate(ColumnSchema columnSchema, KuduP
+            if (type.getClass().getSuperclass().equals(DecimalType.class)) {
+                return KuduPredicate.newComparisonPredicate(columnSchema, op, new BigDecimal((Long) javaValue));
+            } else {
+                return KuduPredicate.newComparisonPredicate(columnSchema, op, (Long) javaValue);
```
We can remove redundant else as below.
```
if () {
 return x;
}
return y;
```

Crossoverrr replied:
Ok, I will do that!

Crossoverrr commented:
thank you for your suggestions, I will try to do that!

ebyhr commented:
@Crossoverrr If you need guidance, you can join the `#dev` channel of our slack https://prestosql.io/slack.html

Crossoverrr commented:
> @Crossoverrr If you need guidance, you can join the `#dev` channel of our slack https://prestosql.io/slack.html

Thank you I will join that channel! lol

Crossoverrr commented:
> hi, Thanks a lot for raising this PR. Can you please squash the two commits into one. I guess you have submitted the CLA

Excuse me, how can I squash the two commits into one, could you help me, please? :)

Crossoverrr commented:
> Thanks for creating this PR!
> 
> Could you add a test to `TestKuduIntegrationSmoke` or update `TestKuduIntegrationDecimalColumns`? The current implementation at L548 will throw the IllegalArgumentException because the scale isn't specified. This is a test example:
> 
> ```java
>     @Test
>     public void testDecimalColumn()
>     {
>         assertUpdate("CREATE TABLE IF NOT EXISTS test_decimal (" +
>                 "id INT WITH (primary_key=true), " +
>                 "col_decimal decimal(10, 6)" +
>                 ") WITH (" +
>                 " partition_by_hash_columns = ARRAY['id'], " +
>                 " partition_by_hash_buckets = 2" +
>                 ")");
> 
>         assertUpdate("INSERT INTO test_decimal VALUES (0, 0.0), (2, 2.2), (1, 1.1)", 3);
>         assertQuery("SELECT * FROM test_decimal WHERE col_decimal = 1.1", "VALUES (1, 1.1)");
> 
>         assertUpdate("DROP TABLE test_decimal");
>     }
> ```
> 
> ```
> java.lang.AssertionError: Execution of 'actual' query failed: SELECT * FROM test_decimal WHERE col_decimal = 1.1
> 
> 	at org.testng.Assert.fail(Assert.java:83)
> 	at io.prestosql.testing.QueryAssertions.assertQuery(QueryAssertions.java:147)
> 	at io.prestosql.testing.QueryAssertions.assertQuery(QueryAssertions.java:103)
> 	at io.prestosql.testing.AbstractTestQueryFramework.assertQuery(AbstractTestQueryFramework.java:135)
> 	at io.prestosql.plugin.kudu.TestKuduIntegrationSmoke.testDecimalColumn(TestKuduIntegrationSmoke.java:123)
> 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
> 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
> 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
> 	at java.lang.reflect.Method.invoke(Method.java:498)
> 	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
> 	at org.testng.internal.Invoker.invokeMethod(Invoker.java:645)
> 	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:851)
> 	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1177)
> 	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129)
> 	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112)
> 	at org.testng.TestRunner.privateRun(TestRunner.java:756)
> 	at org.testng.TestRunner.run(TestRunner.java:610)
> 	at org.testng.SuiteRunner.runTest(SuiteRunner.java:387)
> 	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:382)
> 	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:340)
> 	at org.testng.SuiteRunner.run(SuiteRunner.java:289)
> 	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
> 	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
> 	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1293)
> 	at org.testng.TestNG.runSuitesLocally(TestNG.java:1218)
> 	at org.testng.TestNG.runSuites(TestNG.java:1133)
> 	at org.testng.TestNG.run(TestNG.java:1104)
> 	at org.testng.IDEARemoteTestNG.run(IDEARemoteTestNG.java:73)
> 	at org.testng.RemoteTestNGStarter.main(RemoteTestNGStarter.java:123)
> Caused by: java.lang.RuntimeException: Value precision 13 (after scale coercion) can't be coerced to target precision 10. 
> 	at io.prestosql.testing.AbstractTestingPrestoClient.execute(AbstractTestingPrestoClient.java:114)
> 	at io.prestosql.testing.DistributedQueryRunner.execute(DistributedQueryRunner.java:379)
> 	at io.prestosql.testing.QueryAssertions.assertQuery(QueryAssertions.java:144)
> 	... 27 more
> Caused by: java.lang.IllegalArgumentException: Value precision 13 (after scale coercion) can't be coerced to target precision 10. 
> 	at org.apache.kudu.util.DecimalUtil.coerce(DecimalUtil.java:135)
> 	at org.apache.kudu.client.KuduPredicate.newComparisonPredicate(KuduPredicate.java:253)
> 	at io.prestosql.plugin.kudu.KuduClientSession.createComparisonPredicate(KuduClientSession.java:549)
> 	at io.prestosql.plugin.kudu.KuduClientSession.createEqualsPredicate(KuduClientSession.java:539)
> 	at io.prestosql.plugin.kudu.KuduClientSession.addConstraintPredicates(KuduClientSession.java:493)
> 	at io.prestosql.plugin.kudu.KuduClientSession.buildKuduSplits(KuduClientSession.java:148)
> 	at io.prestosql.plugin.kudu.KuduSplitManager.getSplits(KuduSplitManager.java:49)
> 	at io.prestosql.split.SplitManager.getSplits(SplitManager.java:86)
> 	at io.prestosql.sql.planner.DistributedExecutionPlanner$Visitor.visitScanAndFilter(DistributedExecutionPlanner.java:190)
> 	at io.prestosql.sql.planner.DistributedExecutionPlanner$Visitor.visitFilter(DistributedExecutionPlanner.java:258)
> 	at io.prestosql.sql.planner.DistributedExecutionPlanner$Visitor.visitFilter(DistributedExecutionPlanner.java:150)
> 	at io.prestosql.sql.planner.plan.FilterNode.accept(FilterNode.java:72)
> 	at io.prestosql.sql.planner.DistributedExecutionPlanner.doPlan(DistributedExecutionPlanner.java:120)
> 	at io.prestosql.sql.planner.DistributedExecutionPlanner.doPlan(DistributedExecutionPlanner.java:125)
> 	at io.prestosql.sql.planner.DistributedExecutionPlanner.plan(DistributedExecutionPlanner.java:97)
> 	at io.prestosql.execution.SqlQueryExecution.planDistribution(SqlQueryExecution.java:434)
> 	at io.prestosql.execution.SqlQueryExecution.start(SqlQueryExecution.java:340)
> 	at io.prestosql.$gen.Presto_null__testversion____20191125_135512_3.run(Unknown Source)
> 	at io.prestosql.execution.SqlQueryManager.createQuery(SqlQueryManager.java:240)
> 	at io.prestosql.dispatcher.LocalDispatchQuery.lambda$startExecution$4(LocalDispatchQuery.java:107)
> 	at io.prestosql.$gen.Presto_null__testversion____20191125_135512_3.run(Unknown Source)
> 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
> 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
> 	at java.lang.Thread.run(Thread.java:748)
> ```

I have fixed this problem, please retest, thank you for your help!

ebyhr commented:
@Crossoverrr I sent DM about squashing commits on Slack.

ebyhr commented on presto-kudu/src/main/java/io/prestosql/plugin/kudu/TypeHelper.java:
```diff
@@ -156,7 +158,9 @@ public static Object getJavaValue(Type type, Object nativeValue)
         if (type instanceof DecimalType) {
-            return nativeValue;
+            DecimalType dt = (DecimalType) type;
+            return new BigDecimal(BigInteger.valueOf((Long) nativeValue),
```
`(Long) nativeValue` will throw ClassCastException if the value is larger than max value of long. We can avoid the situation by;
```java
            DecimalType decimalType = (DecimalType) type;
            if (decimalType.isShort()) {
                return new BigDecimal(BigInteger.valueOf((long) nativeValue), decimalType.getScale());
            }
            return new BigDecimal(decodeUnscaledValue((Slice) nativeValue), decimalType.getScale());
```

tips: We don't use abbreviations (`dt`)

Crossoverrr replied:
Thank you for your help, I really appreciate that! [handshake]

Crossoverrr commented:
> hi, Thanks a lot for raising this PR. Can you please squash the two commits into one. I guess you have submitted the CLA

Oh, thank you for your remind! I've mailed my CLA to cla@prestosql.io yesterday, but I found my verification/cla-signed still failed today.  :(

martint commented:
@cla-bot check

Praveen2112 commented:
Can we do a rebase instead of doing a merge with master ?

Crossoverrr commented:
Sorry I think I just messed up this pr , I think I should close this one, and recreate a new one.
